contract/src/Custody.sol

```diff
@@ -369,41 +362,42 @@ contract Custody is IChannel, IDeposit {
      */
-    // TODO: add responding to CHANOPEN, CHANRESIZE challenge (should NOT call `adjudicate`)
     function checkpoint(bytes32 channelId, State calldata candidate, State[] calldata proofs) external {
         Metadata storage meta = _channels[channelId];
 
-        // Verify channel exists and is not VOID or FINAL
+        // checks
         if (meta.stage == ChannelStatus.VOID) revert ChannelNotFound(channelId);
         if (meta.stage == ChannelStatus.FINAL) revert InvalidStatus();
 
-        // Verify that at least one participant signed the state
-        if (candidate.sigs.length == 0) revert InvalidStateSignatures();
-
-        // Validate version based on channel status
-        if (meta.stage == ChannelStatus.INITIAL && candidate.version != 0) revert InvalidState();
-
-        if (!IAdjudicator(meta.chan.adjudicator).adjudicate(meta.chan, candidate, proofs)) revert InvalidState();
-
-        // Verify this state is more recent than the current stored state
-        if (candidate.version == meta.lastValidState.version) {
-            // Explicitly check for equal versions and reject them
+        // if INITIALIZE, call `join(...)`. If RESIZE, call `resize(...)`. If FINALIZE, call `close(...)`.
+        if (candidate.intent != StateIntent.OPERATE) {
             revert InvalidState();
         }
 
-        if (!_isMoreRecent(meta.chan.adjudicator, candidate, meta.lastValidState)) {
-            revert InvalidState();
-        }
+        StateIntent lastValidStateIntent = meta.lastValidState.intent;
 
-        // Store the candidate as the last valid state
-        meta.lastValidState = candidate;
+        // main goal: verify Candidate is valid and > LastValidState
+
+        if (meta.stage == ChannelStatus.INITIAL) {
+            revert InvalidStatus(); // Cannot checkpoint in INITIAL stage, use `join(...)` instead
+        } else if (meta.stage == ChannelStatus.ACTIVE) {
+            if (!_isMoreRecent(meta.chan.adjudicator, candidate, meta.lastValidState)) revert InvalidState();
+            if (!IAdjudicator(meta.chan.adjudicator).adjudicate(meta.chan, candidate, proofs)) revert InvalidState();
+        } else {
+            // meta.stage == ChannelStatus.DISPUTE
+            if (!IAdjudicator(meta.chan.adjudicator).adjudicate(meta.chan, candidate, proofs)) revert InvalidState();
+
+            if (lastValidStateIntent == StateIntent.OPERATE) {
+                if (!_isMoreRecent(meta.chan.adjudicator, candidate, meta.lastValidState)) revert InvalidState();
+            }
 
-        // If there's an ongoing challenge and this state is newer, cancel the challenge
-        if (meta.stage == ChannelStatus.DISPUTE) {
-            meta.stage = ChannelStatus.ACTIVE;
             meta.challengeExpire = 0;
         }
 
+        // effects
+        meta.stage = ChannelStatus.ACTIVE;
+        meta.lastValidState = candidate;
+
         emit Checkpointed(channelId);
```

Comment:
Maybe add turn number (and perhaps state hash) to the Checkpointed and Challenged events? It should be useful

Reply:
4f0369c385e79a678e5ec41b16271652a7d944b9